crossplane/provider-gcp
# Add Nodepool v1alpha1 type back and set v1beta1 as storage version

### Description of your changes


_**Adding only type, no controller.**_

Currently, when we want to upgrade `provider-gcp` from `v0.17.x` to the latest master (i.e. release candidate for `v0.18.0`), it ends up with an **unhealthy providerrevision** with the following error:

```
cannot establish control of object: CustomResourceDefinition.apiextensions.k8s.io "nodepools.container.gcp.crossplane.io" is invalid: status.storedVersions[0]: Invalid value: "v1alpha1": must appear in spec.versions
``` 

We have removed `v1alpha1` version for `NodePool` resource as part of [removing dependency pins to gcloud beta sdk work](https://github.com/crossplane/provider-gcp/pull/308), and planning to provide a migration path to the new NodePool version. However, removing `v1alpha1` version for `NodePool` removed it from CRD definition resulting the above error. Please note, you can observe the same error independent of crossplane packaging, by first creating NodePool CRD from `release-0.17` and then applying from `master`.

This PR fixes this problem by re-adding `v1alpha1` version for `NodePool` (as it is in `v0.17.1`) and setting `v1beta1` as the storage version. This is also consistent with the migration path suggested in k8s docs [here](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#upgrade-existing-objects-to-a-new-stored-version).

This PR also adds a deprecation warning for `v1alpha1` version (**which required to bump controller-tools version causing trivial changes on all CRDs**):

```
$ kubectl get nodepools.v1alpha1.container.gcp.crossplane.io

Warning: container.gcp.crossplane.io/v1alpha1 NodePool is deprecated; please use v1beta1 version instead
NAME                 READY   SYNCED   STATE   CLUSTER-REF       AGE
crossplane-np                False            example-cluster   69m
crossplane-np-beta           False            example-cluster   64m
```

Please note, this is not an issue for `GKECluster` since it is renamed as `Cluster`, which means the new revision takes no action on the old CRD.

I have:


### How has this code been tested

1. Try manual CRD apply: 

    ```
    git checkout v0.17.1
    kubectl create -f package/crds/container.gcp.crossplane.io_nodepools.yaml
    git checkout keep-old-nodepool # this branch
    kubectl apply -f package/crds/container.gcp.crossplane.io_nodepools.yaml
    ```
2. Try upgrading with Crossplane package manager:

    ```
    kubectl crossplane install provider crossplane/provider-gcp:v0.17.1
    kubectl edit provider.pkg.crossplane.io/crossplane-provider-gcp # and set package as turkenh/provider-gcp-controller:keep-old-nodepool-v2
    ```

Also tested by creating example Nodepool resource from v1alpha1 and v1beta1 and verifying the behavior.

[contribution process]: https://git.io/fj2m9
